File: packages/sdk-typescript/src/instrumentation/anthropic.ts
import type { Tracer } from '@opentelemetry/api'
import { SpanStatusCode } from '@opentelemetry/api'
import shimmer from 'shimmer'

type AnthropicMessagesCreate = (params: {
  model: string
  messages: unknown[]
  max_tokens: number
  [key: string]: unknown
}) => Promise<{
  model: string
  usage: { input_tokens: number; output_tokens: number } | null | undefined
  [key: string]: unknown
}>

type AnthropicLike = {
  messages: { create: AnthropicMessagesCreate }
}

export function patchAnthropic(tracer: Tracer, client: AnthropicLike | undefined): void {
  if (!client) return

  shimmer.wrap(client.messages, 'create', (original: AnthropicMessagesCreate) => {
    return async function patchedCreate(
      this: unknown,
      params: Parameters<AnthropicMessagesCreate>[0]
    ): Promise<Awaited<ReturnType<AnthropicMessagesCreate>>> {
      const span = tracer.startSpan('anthropic.messages.create')
      span.setAttribute('tracion.kind', 'llm')
      span.setAttribute('llm.provider', 'anthropic')
      span.setAttribute('llm.model', params.model)

      try {
        const result = await original.call(this, params)
        if (result.usage) {
          span.setAttribute('llm.input_tokens', result.usage.input_tokens)
          span.setAttribute('llm.output_tokens', result.usage.output_tokens)
        }
        span.setStatus({ code: SpanStatusCode.OK })
        span.end()
        return result
      } catch (e) {
        span.setStatus({
          code: SpanStatusCode.ERROR,
          message: e instanceof Error ? e.message : String(e),
        })
        span.end()
        throw e
      }
    }
  })
}

/** Anthropic SDK インスタンスを自動検出してパッチする */
export function tryPatchAnthropic(tracer: Tracer): void {
  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const { Anthropic } = require('@anthropic-ai/sdk') as { Anthropic: new () => AnthropicLike }
    const proto = Anthropic.prototype as unknown as AnthropicLike
    patchAnthropic(tracer, proto)
  } catch {
    // @anthropic-ai/sdk がインストールされていない場合はスキップ
  }
}

File: packages/sdk-python/src/tracion/instrumentation/_anthropic.py
from typing import Any

from opentelemetry.trace import Tracer, StatusCode, Status


def patch_anthropic(tracer: Tracer, client: Any) -> None:
    """Anthropic クライアントの messages.create をパッチしてトークンを記録する。

    client が None の場合は何もしない（anthropic 未インストール時のフォールバック）。
    """
    if client is None:
        return

    original_create = client.messages.create

    def patched_create(*args: Any, **kwargs: Any) -> Any:
        span = tracer.start_span("anthropic.messages.create")
        span.set_attribute("tracion.kind", "llm")
        span.set_attribute("llm.provider", "anthropic")

        model = kwargs.get("model") or (args[0] if args else None)
        if model:
            span.set_attribute("llm.model", str(model))

        try:
            result = original_create(*args, **kwargs)
            usage = getattr(result, "usage", None)
            if usage:
                input_tokens = getattr(usage, "input_tokens", None)
                output_tokens = getattr(usage, "output_tokens", None)
                if input_tokens is not None:
                    span.set_attribute("llm.input_tokens", int(input_tokens))
                if output_tokens is not None:
                    span.set_attribute("llm.output_tokens", int(output_tokens))
            span.set_status(Status(StatusCode.OK))
            span.end()
            return result
        except Exception as e:
            span.set_status(Status(StatusCode.ERROR, str(e)))
            span.end()
            raise

    client.messages.create = patched_create


def try_patch_anthropic(tracer: Tracer) -> None:
    """Anthropic SDK が存在する場合に自動パッチを試みる。"""
    try:
        import anthropic  # noqa: PLC0415
        patch_anthropic(tracer, anthropic.Anthropic)
    except ImportError:
        pass
